extensions/amp-list/0.1/amp-list.js

```diff
@@ -220,15 +220,27 @@ export class AmpList extends AMP.BaseElement {
 
   /**
    * Dequeues a fetch result and renders it immediately.
+   * Queues another render pass when render completes.
    * @private
    */
   doRenderPass_() {
-    dev().assert(this.renderQueue_.length > 0, 'Nothing queued to render.');
+    dev().assert(this.renderQueue_.length > 0, 'Nothing queued for render.');
     const {items, resolver, rejecter} = this.renderQueue_.shift();
+    const scheduleNextPass = () => {
+      if (this.renderQueue_.length) {
+        this.renderPass_.schedule(1); // Allow paint frame before next render.
```

Comment:
I'm curious why we need multiple renders to begin with? Why not just store the latest items (only), schedule pass, and render it?

Reply:
That'd work too, a queue was more intuitive to me. In practice it doesn't make a difference since user gestures can't trigger that quickly anyways.

Reply:
I'm not sure I agree.

```js
class AmpList {
  constructor() {
    // ...
    
    this.renderItems_ = [];
    this.renderPromise_ = null;
    this.renderPromiseResolve_ = null;
  }

  scheduleRender_(items) {
    if (!this.renderPromise_) {
      this.renderPromise_ = new Promise(resolve => {
        this.renderPromiseResolve_ = resolve;
      });
    }
    this.renderItems_ = items;
    this.renderPass_.schedule();
    return this.renderPromise_;
  }

  doRenderPass_() {
    const render = this.templates_.findAndRenderTemplateArray(this.element, items)
        .then(elements => this.updateBindings_(elements))
        .then(elements => this.rendered_(elements));
    this.renderPromiseResolve_(render);
    this.renderPromise_ = null;
    this.renderPromiseResolve_ = null;
  }

Reply:
This also needs to check if `this.renderItems_` has changed by the time the render completes and schedule another pass if so. I'll try removing the queue and see if it looks cleaner.

Reply:
> has changed by the time the render completes and schedule another pass if so.

Why? Can't it just render with the latest items?

Reply:
The whole `doRenderPass_` process takes awhile (several frames) and in that time another mutation can be triggered. This caused a race condition due to interleaving of render sub-processes like `updateBindings_` which was the bug in the original issue.

Reply:
> The whole doRenderPass_ process takes awhile (several frames) and in that time another mutation can be triggered. 

This is a job for cancellation tokens.

```js
class AbortSignal {
  constructor() {
    this.isAborted_ = false;
  }

  abort() {
    this.isAborted_ = true;
  }

  throwIfAborted() {
    if (this.isAborted_) {
      throw new Error('...');
    }
  }
}



  doRenderPass_() {
    const signal = new AbortSignal();
    this.signal_ = signal;
    const render = this.templates_.findAndRenderTemplateArray(this.element, items)
        .then(elements => this.updateBindings_(elements))
        .then(elements => this.rendered_(elements));
    this.renderPromiseResolve_(render);
    this.renderPromise_ = null;
    this.renderPromiseResolve_ = null;
  }

  updateBindings_(elements) {
    this.signal_.throwIfAborted();
    // not aborted, proceed
  }

  render_(elements) {
    this.signal_.throwIfAborted();
    // not aborted, proceed
  }
```

> This caused a race condition due to interleaving of render sub-processes like updateBindings_ which was the bug in the original issue.

This should still be fixed by my code, since we're syncly removing the promise. Any mutation that comes in after the start of the last `doRenderPass_` will schedule their own pass.

Reply:
- I did consider canceling the pass mid-flight but I don't think it's a necessary optimization and I wanted to keep things as simple as possible.
- I don't think that'll prevent the race -- let's chat more in person. 
- I did replace the queue with a single reference to the latest fetch results -- much better since it avoids wasting work. 👍 

Thanks for the early approval. Tested these changes and going to merge soon to make this week's cut.